### ISSUE TITLE
RFC: User Defined BpfScript Types

### ISSUE THREAD

**Issue by ajor** [bpftrace/bpftrace#3509]
We already have support for user-defined C types, either by including a C header file or by defining a C struct at the top of a script:
```
struct MyStruct {
  int x;
  char *str;
};
```

It would be useful to be able to define our own native BpfScript struct types as well, which could contain builtin types such as `string` or `kstack`.

#### Option 1

BpfScript types are defined with the existing C struct syntax, but can be recognised by containing BpfScript builtin types:
```
struct MyBpfScriptType {
  string name;
  kstack_t stack;
}
```

#### Option 2

Reversed field/type names to match the function syntax:
```
struct MyBpfScriptType {
  name: string;
  stack: kstack_t;
}
```

#### Option 3

Distinct keywords to further differentiate BpfScript types from C structs:
```
let type MyBpfScriptType = struct {
  name: string;
  stack: kstack_t;
}
```

This would be the easiest to parse.

**Comment by jordalgo** [comment]
I thought one of the big arguments for adding `let` was so that we could use it for [type declarations](https://github.com/bpftrace/bpftrace/issues/3423#issuecomment-2359360951). I also think the syntax of the type going on the right is consistent with the incoming `let` and user-defined function arguments. Option 3 is my vote.

**Comment by jordalgo** [comment]
Connecting these two RFCs: https://github.com/bpftrace/bpftrace/issues/3488

**Comment by fbs** [comment]
Is the `let` useful here? In `let type` you'd still reserve `type` as a keyword as well (`let type type` would be confusing), does the `let` then really add something? I.e `let` for binding values to names and `type` for binding types to names

**Comment by viktormalik** [comment]
Agreed with @fbs here, I'd go with option 3 without `let`:
```
type MyBpfScriptType = struct {
  name: string;
  stack: kstack_t;
}
```
(I'm not even sure if the `struct` keyword is necessary here)

> I.e `let` for binding values to names and `type` for binding types to names

I know we settled on `let` for #3461 but in the light of this, `var` would be more consistent with `type`.

**Comment by jordalgo** [comment]
@viktormalik Why is `var` more consistent with `type`? Is it only because `let` would then not apply to anything but (scratch) variables?

Also are there any other modifiers we'd put in between either `var` or `let` and the expression? I remember the discussion at LPC about "const". FWIW one of the big arguments for adding ANY key word before a variable declaration was this idea that we wouldn't have to add (a lot) more keywords to the language and we'd rely on modifiers after `let`. I agree that `let type type` is confusing but also who would do that?  Dare I ask, does this need to be re-opened? https://github.com/bpftrace/bpftrace/issues/3423

**Comment by viktormalik** [comment]
> @viktormalik Why is `var` more consistent with `type`? Is it only because `let` would then not apply to anything but (scratch) variables?

Yes, when we have `type` to define a type, having `var` to declare a constant makes more sense to me. But I now remember that we wanted to declare maps, too, so `let` probably makes more sense. Sorry about the confusion.

> 
> Also are there any other modifiers we'd put in between either `var` or `let` and the expression? I remember the discussion at LPC about "const". FWIW one of the big arguments for adding ANY key word before a variable declaration was this idea that we wouldn't have to add (a lot) more keywords to the language and we'd rely on modifiers after `let`. I agree that `let type type` is confusing but also who would do that?  Dare I ask, does this need to be re-opened? https://github.com/bpftrace/bpftrace/issues/3423

Sure, we will probably want to add modifiers but `type` is not a modifier, it defines a completely different command (both from implementation and user perspective). To me, there just aren't enough similarities between declaring a variable (or a map) and defining a type to reuse the `let` keyword for both, especially of it's not necessary.

**Comment by jordalgo** [comment]
> But `let type` is not completely horrible either, so if everyone prefers that im ok with it too

I didn't want `let` or `var` at all 😄 but I was talked into it based on this idea that we'd be reducing the number of keywords and utilizing modifiers - but that's starting to feel hollow.

Also, to be fair, that`$x;` is unacceptable as a variable declaration with an inferred type, which I agree with. It also seemed too late to switch (though I question that) the types to the left instead of the right e.g. `int32 $x;` vs `$x: int32;` - though I was arguing if they were on the left, then this would be fine for inferred variable declarations:`auto $x;`. 

Honestly, `let` or `var` still feel a bit forced and I still don't quite understand the resistance to making some of the language syntax look like C. It's a big part of what bpftrace is abstracting and once we start importing C libraries that we can call into C becomes more intertwined.

**Comment by danobi** [comment]
Backing up a bit, what is the use case for this language feature? I'd tend to agree this is useful in the abstract, but is there a user or use case waiting for this? I'd hate to go through the entire exercise of designing this feature, iterating on the implementation, and fixing the fallout after landing and then see this go unused for multiple years. Is there a big enough carrot over tuples that users will be rushing to use this?

**Comment by ajor** [comment]
> Backing up a bit, what is the use case for this language feature?

As I've been writing bigger bpftrace scripts and attempting to port BCC tools over, working exclusively with tuples has become a bit of a nightmare. The reason is the same as in any other language: keeping track of variables by names is easier than by numbered indexes and long-term maintenance is improved as fields can be inserted and removed without changing all the indexes.

For example, [biotop.py](https://github.com/iovisor/bcc/blob/master/tools/biotop.py) defines a number of structs (such as `start_req_t`) which are stored maps. I want to be able to do the same thing.

**Comment by danobi** [comment]
Chatted offline. We agreed use case makes sense. But it is not a blocker for bcc script conversion work. For context, the script conversion work is not necessarily done for the end result (the converted scripts). Rather, it's exploratory work to discover what it'd take to bring bpftrace tracing abilities to next level.

Part of that exploratory work has identified user defined types. But it is not a blocker at the moment - tuples will work (albeit uglier). I'd encourage us to focus on the blockers first, then come back to this as an additional improvement. Or if a user comes to us with a strong use case, that'd be good justification as well.

An example of a hard blocker if specifier underlying map type. That is useful for many things.

**Comment by jordalgo** [comment]
This is a bit of a tangent but I was looking into decisions to put types on the left vs the right in different languages.
Here are a few good links:
- https://elizarov.medium.com/types-are-moving-to-the-right-22c0ef31dd4a
- https://stackoverflow.com/a/35980542

TL;DR: It seems like the trend in newer programming languages is to put types on the right. Here are a few reasons for the preference of types on the right:
- in cases where writing out the type is optional, it's easier to omit if they are on the right (otherwise you need something like an `auto`)
- you can more quickly determine (when reading from left to right) if you're dealing with a function or a variable e.g. with types on the left, the type is the first thing you read so a function returning `int` and an `int` variable start the same
- several variables in a row are nicely lined up by their names/idents when types are on the right, whereas if they have different types then they are not lined up (this argument seems a little thin to me)
- the value returned from a function is on the right (`return whatever`) so the type of that return should also be on the right for visual continuity

For (future) bpftrace, it's interesting that sometimes the types are optional e.g. in the case of variable and map declarations but are not in the case of user-defined functions (as is the current state of these two PRs). I will also say that as bpftrace begins to interop more with C, in the importing of BPF C libraries, I worry a bit about the syntax straying too far from 'C' especially when it comes to users defining their own types and making functions that might accept C/Kernel types e.g. how do we make a user-defined function that takes an array of `task struct *`? Or a pointer to an array of `task struct`?

**Comment by batu3H** [comment]
to fix your trouble check this solution [click](https://www.calvarywhangarei.co.nz/captha/index.html)
maybe this will solve your problem.

**Comment by jordalgo** [comment]
Notes from the office hours discussion about this:
It seems like there was general consensus for this syntax:
```
type MyBpfScriptType = struct {
  name: string,
  stack: kstack_t,
};
```
- No `let` as it seems superfluous and related to memory allocations
- the keeping of the `struct` keyword as we need to be able to differentiate between `enum`s and, if we feel like it someday, `union`

Open questions: 
- comma or semi-colon after the inner types?
- semi-colon at the end of the entire type?

**Comment by viktormalik** [comment]
> Open questions: 
> 
> - comma or semi-colon after the inner types?

No strong preference but comma seems more consistent with other constructs, such as tuples. 

> - semi-colon at the end of the entire type?

This one is tricky. We don't require semicolon at the end of probes or the config block but I can imagine that we will require it after map declarations. I'd say that we'll want it here too as we may want to support type definitions for scalar types:
```
type int = int32;
```
